addons/mass_mailing/models/mailing.py

```diff
@@ -260,15 +260,34 @@ def _compute_statistics(self):
             self.browse(row.pop('mailing_id')).update(row)
 
     def _compute_next_departure(self):
-        cron_next_call = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo().nextcall
+        cron_job = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo()
         str2dt = fields.Datetime.from_string
-        cron_time = str2dt(cron_next_call)
+        cron_time = str2dt(cron_job.nextcall)
+        cron_id = cron_job.id
+
+        # A cron.trigger is created when mailing is put "in queue"
+        # The cron workers usually clean up old cron.triggers after finishing their
+        # jobs, i.e older than now(). Thus only most recent than now is relevant
+        now_datetime_buffer = fields.Datetime.now() + relativedelta(minutes=1)
+        cron_triggers = self.env['ir.cron.trigger'].sudo().search([
+            '&',
+            ('cron_id', '=', cron_id),
+            ('call_at', '<', now_datetime_buffer)],
+            order='call_at desc', limit=1)
+        most_recent_trigger_datetime = cron_triggers.call_at
         for mass_mailing in self:
             if mass_mailing.schedule_date:
-                schedule_date = str2dt(mass_mailing.schedule_date)
-                mass_mailing.next_departure = max(schedule_date, cron_time)
+                mass_mailing.next_departure = max(mass_mailing.schedule_date, fields.datetime.now())
             else:
-                mass_mailing.next_departure = cron_time
+                # Schedule_date should is only False if schedule_type = "now" or
+                # mass_mailing is canceled.
+                # We checked for cron.triggers relating to the cron_id
+                # so we can reasonably expect that the cron worker will
+                # execute this mass_mailing job the next time it wakes up
+                if most_recent_trigger_datetime:
+                    mass_mailing.next_departure = most_recent_trigger_datetime
+                else:
+                    mass_mailing.next_departure = cron_time
```

Comment:
Simpliefied code based on feedback.

Concerning `cron_time`: as we now rely on cron trigger times, I guess we could get rid of that "fail safe"? Normally based on current code behavior, we should always have at least one trigger created thus outputting either the schedule_date/trigger_datetime or `now`, whichever is more relevant. 

Let me know if we should remove it, as I mostly left it based on previous authors' commit and in case I was missing some unforeseen edge case.

Reply:
I think having a fail safe is always a good idea :) we never know what people did with their data ... does not cost much.